### ISSUE TITLE
Is this a reasonable approach?

### ISSUE THREAD

**Issue by mjuniper** [mjuniper/ember-facebook-video-demo#1]
I created a [component](https://github.com/mjuniper/ember-facebook-video-demo/blob/master/app/components/facebook-video.js) for a facebook video embed. Usage [here](https://github.com/mjuniper/ember-facebook-video-demo/blob/master/app/templates/application.hbs#L1-L11).

It loads the facebook sdk in the component's `init` method (if it has not already been loaded). That way it only gets loaded if a site/page uses this component and it will only be loaded once.

I have some questions/reservations about this:

1. Is this a good approach generally / does anybody have a better idea?
2. Will this be fastboot friendly (I assume the use of `document` means the answer is no).
3. Facebook's video embed [instructions](https://developers.facebook.com/docs/plugins/embedded-video-player/#configurator) indicate that you should include an appId. But it works without it. I think it might only say that because my facebook account has an app registered to it. Not sure if this is an issue.

**Comment by MarvinPerry** [comment]
Reading the the Ember guides, and they suggest `didInsertElement` as a [good entry point](https://guides.emberjs.com/release/components/the-component-lifecycle/#toc_integrating-with-third-party-libraries-with-didinsertelement) for bringing in 3rd party js. I know tom has done a lot of work to shrink our js footprints, and I think doing it at the component level would also keep us from loading it needlessly.  Twitter will also require us to take their js this way. 

Another method I've seen recommended is to use jquery's `getScript` [method](https://api.jquery.com/jquery.getscript/) to pull in their js.  I'm not entirely sure this method doesn't go against the Ember way though. I feel like it might also cause weird component reload issues.

**Comment by mjuniper** [comment]
Just to be clear - this _does_ load the js in the component. It just uses the `init` lifecycle hook instead of `didInsertElement`. I think `init` is better than `didInsertElement` for this sort of thing because it happens earlier. Also loading this script doesn't need the component to have DOM so I don't see a reason to wait until `didInsertElement`. But it probably doesn't matter that much, `didInsertElement` would work just fine.

We want to avoid jquery when we can.

**Comment by MarvinPerry** [comment]
Right, that makes a lot of sense. My concern with `init` was if they updated the video to render. I didn't think `init` would get triggered again unless they rendered a new component.  I see now that 1. If that was  going to be a problem, it'd also be the case for `didInsertElement` and 2. it's probably not an issue since we're talking about when we make that javascript available, not when we use it to render something. 

I'm not sure if I'm explaining myself well, but yeah, I think `init` is good.

**Comment by mjuniper** [comment]
Yes that is correct on both counts.

**Comment by tomwayson** [comment]
Generally speaking, using `document.createElement(script)` is the best way that I've found to lazy-load dependencies like this, since dynamic `import()` is not ready yet, especially in Ember. That's how [esri-loader works](https://github.com/Esri/esri-loader/blob/741ba5f10ef42373e7ac6308e2875eb0261b0002/src/esri-loader.ts#L20-L26) and how we [lazy-load amCharts](https://github.com/Esri/ember-cli-cedar/blob/0a14ec0302e2c3c95d6e7b8122349b82faca95dd/addon/utils/lazy-loading.js#L4-L11). The main difference w/ those is that instead of checking for the existence of the `<script>` tag by looking for it's `id` they check for the existence of globals set by the libraries they load (`require()` and `AmCharts` respectively). That allows them to work if the library was loaded some other way. I think in your case using the `id` is fine b/c this is an app not a library and we can avoid `id` collisions and we don't expect the FB lib to be loaded in any other way than via this component.

You should definitely use `didInsertElement` instead of `init()`. The main reason is that `didInsertElement()` is **not** called in Fastboot, so you can safely use `document`. That's why I use `didInsertElement()` [in ember-esri-loader](https://github.com/Esri/ember-esri-loader#loading-modules-from-the-arcgis-api-for-javascript) and ember-cli-cedar.

**Comment by mjuniper** [comment]
Awesome, thx @tomwayson. Didn't know that about fastboot/didInsertElement.
